lib/puppet/module.rb

```diff
@@ -90,7 +90,7 @@ def has_metadata?
     return false unless Puppet::FileSystem.exist?(metadata_file)
 
     begin
-      metadata =  JSON.parse(File.read(metadata_file))
+      metadata =  JSON.parse(File.read(metadata_file, :encoding => 'utf-8'))
```

Comment:
Fair point and thanks for bringing it up.  A few counterpoints / ideas:

* JSON is specified as Unicode in [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt), with the default being UTF-8.  The Encoding section of the spec basically states "look at the BOM" to be certain, but only declares UTF-8, UTF-16LE, UTF-16BE, UTF32-LE and UTF32-BE as valid encodings.
* That said, it doesn't mean users creating files with a text editor are necessarily conforming to the spec.
* From a pragmatic standpoint, I think the likelihood that there are characters in metadata using other encodings, that can't parse as UTF-8 is unlikely. I'm not sure how likely it might be to corrupt the payload - for instance:
  * with codepages that might trigger a multibyte character sequence in UTF-8, thereby throwing off the rest of the file parsing
  * with codepages using completely invalid UTF-8 byte sequences (there are some tix in JIRA where this happened trying to parse control sequences from linux command line tool output)
  * codepages that have characters like `"`, `{` or other crucial JSON constructs at different positions than ASCII or UTF-8
  * that is, I believe it's more common to see `Invalid Byte Sequence` problems when going to `US-ASCII` or from `UTF-16LE` / `UTF-8` to a local codepage
* I think the biggest concern IMHO is if the file was written on Windows as `UCS-2` and ends up being read as `UTF-8`, which typically introduces characters spaced between NULLs due to the extra bytes - which should show up as a parsing error when trying to treat it as JSON.
* I don't think this is actually a behavioral change for the master (which *should* be using UTF-8 by default anyhow), but it is a change for agents using PMT I believe.  In other words, this makes the `.read` explicit - but I don't believe Ruby would have parsed `UCS-2` correctly previously anyhow.  We can certainly validate that to make 100% sure.

So I think your point about being more defensive might be a good one, since we have no control over module payloads... though I wonder if the best spot to nip problems in the bud is during a `module build` when the `.tar.gz` is produced ... or upon upload to the forge (I think that `module validate` already does this, but not 100% certain).

So I think we can do 2 things to test this a bit more:
* Create a module with a `UCS-2` metadata.json and install it on masters / agents prior to this change / with this change.. to verify.
* Systematically install every module from the forge given the local forge backup ... and see what happens.